program/src/processor/slash.rs

```diff
@@ -11,14 +27,257 @@ use crate::instruction::accounts::{Context, SlashAccounts};
+        ctx.accounts.stake.key == &derivation,
+        ProgramError::InvalidSeeds,
+        "stake",
+    );
+
+    // slash authority
+    // - must be a signer
+    // - must match the slash authority on the config account
+    //
+    // When there is no slash authority set, the stake account cannot be slashed and
+    // an error is returned.
+
+    if let Some(slash_authority) = Option::<Pubkey>::from(config.slash_authority) {
+        require!(
+            ctx.accounts.slash_authority.key == &slash_authority,
+            StakeError::InvalidAuthority,
+            "slash authority",
+        );
+    } else {
+        return err!(StakeError::AuthorityNotSet, "slash authority");
+    }
+
+    require!(
+        ctx.accounts.slash_authority.is_signer,
+        ProgramError::MissingRequiredSignature,
+        "stake authority",
+    );
+
+    // vault authority
+    // - derivation must match
+
+    let signer_bump = [config.vault_authority_bump];
+    let derivation = create_vault_pda(ctx.accounts.config.key, &signer_bump, program_id)?;
+
+    require!(
+        ctx.accounts.vault_authority.key == &derivation,
+        StakeError::InvalidAuthority,
+        "vault authority",
+    );
+
+    let signer_seeds = get_vault_pda_signer_seeds(ctx.accounts.config.key, &signer_bump);
+
+    // vault
+    // - must be the token account on the stake config account
+
+    require!(
+        ctx.accounts.vault.key == &config.vault,
+        StakeError::IncorrectVaultAccount,
+    );
+
+    let vault_data = ctx.accounts.vault.try_borrow_data()?;
+    // unpack to validate the mint
+    let vault = PodStateWithExtensions::<PodAccount>::unpack(&vault_data)?;
+
+    // mint
+    // - must match the stake vault mint
+
+    require!(
+        &vault.base.mint == ctx.accounts.mint.key,
+        StakeError::InvalidMint,
+        "mint"
+    );
+
+    // Update the stake amount on both stake and config accounts:
+    //
+    //   1. the amount slashed is taken from the active amount first;
+    //
+    //   2. in case the amount is not enough, the remaining is taken from
+    //      the inactive amount;
+    //
+    //   3. if the inactive stake is not enough, the remaining is taken
+    //      from the deactivating amount;
+    //
+    //   4. if still not enough, the remaining is ignored and the stake
+    //      account is left with 0 amount.
+
+    require!(
+        amount > 0,
+        StakeError::InvalidAmount,
+        "amount must be greater than 0"
+    );
+
+    let mut remaining_slash_amount = amount;
+
+    // First slash active stake.
+    let active_stake_to_slash = min(amount, stake.amount);
+    stake.amount = stake
+        .amount
+        .checked_sub(active_stake_to_slash)
+        .ok_or(ProgramError::ArithmeticOverflow)?;
+    remaining_slash_amount = remaining_slash_amount.saturating_sub(active_stake_to_slash);
+
+    // Then, if the slash amount has not been fulfilled, slash inactive stake
+    // (if there are inactive tokens).
```

Comment:
Normally, inactive tokens shouldn't be touchable by slashing -- the whole point is that they're inactive! But we can double-check with Uri about this to be totally sure.